File: libs/js/gaming_billing.js
//@ts-check
import axios from "axios";
import { subtle } from "node:crypto";
import Qs from "qs";

export default class GamingBillingAPI {
  constructor(
    endpoint,
    serviceName,
    secretKey,
    serviceHeader = "X-SERVICE",
    signatureHeader = "X-SIGNATURE",
    timestampHeader = "X-SIGNATURE-TIMESTAMP"
  ) {
    this.endpoint = endpoint;
    this.serviceName = serviceName;
    this.serviceHeader = serviceHeader;
    this.signatureHeader = signatureHeader;
    this.timestampHeader = timestampHeader;

    this.client = axios.create({ baseURL: this.endpoint });

    this._secret_key = secretKey;
    this._importedKey = null;

    this._encoder = new TextEncoder();
  }

  async _importKey() {
    this._importedKey = await subtle.importKey(
      "raw",
      this._encoder.encode(this._secret_key),
      { name: "HMAC", hash: "SHA-256" },
      false,
      ["sign"]
    );

    this._secret_key = "";
  }

  async _computeSignatureForData(data) {
    if (this._importedKey === null) {
      await this._importKey();
    }

    //@ts-expect-error _importedKey is defined
    return Buffer.from(await subtle.sign("HMAC", this._importedKey, this._encoder.encode(data))).toString("hex");
  }

  async _getHeaders(path, data) {
    let timestamp = new Date().toISOString();

    let signature;
    if (data === undefined) {
      signature = await this._computeSignatureForData(`${timestamp}.${path}.`);
    } else {
      signature = await this._computeSignatureForData(`${timestamp}.${path}.${data}`);
    }

    return {
      [this.serviceHeader]: this.serviceName,
      [this.signatureHeader]: signature,
      [this.timestampHeader]: timestamp,
      "Content-Type": "application/json",
    };
  }

  async holdersList(filters = {}) {
    let path = "/api/currencies/holders/?" + Qs.stringify(filters, { arrayFormat: "comma" });

    let response = await this.client.get(path, {
      headers: await this._getHeaders(path),
    });

    return response;
  }

  async holdersDetail(holder_id) {
    let path = `/api/currencies/holders/detail/?holder_id=${holder_id}`;

    let response = await this.client.get(path, {
      headers: await this._getHeaders(path),
    });

    return response;
  }

  async holdersCreate(holder_id, holder_type, info = {}) {
    let path = `/api/currencies/holders/create/`;
    let data = JSON.stringify({ holder_id: holder_id, holder_type: holder_type, info: info });

    let response = await this.client.post(path, data, {
      headers: await this._getHeaders(path, data),
    });

    return response;
  }

  async holdersUpdate(holder_id, { enabled, info }) {
    if (enabled === undefined && info === undefined) {
      throw new Error("No parameters are set");
    }

    let path = `/api/currencies/holders/update/`;
    let data = JSON.stringify({ holder_id: holder_id, enabled: enabled, info: info });

    let response = await this.client.post(path, data, {
      headers: await this._getHeaders(path, data),
    });

    return response;
  }

  async accountsList(filters = {}) {
    let path = "/api/currencies/accounts/?" + Qs.stringify(filters, { arrayFormat: "comma" });

    let response = await this.client.get(path, {
      headers: await this._getHeaders(path),
    });

    return response;
  }

  async accountsDetail(holder_id, unit_symbol, holder_type) {
    let path =
      "/api/currencies/accounts/detail/?" +
      Qs.stringify(
        { holder_id: holder_id, unit_symbol: unit_symbol, holder_type: holder_type },
        { arrayFormat: "comma" }
      );

    let response = await this.client.get(path, {
      headers: await this._getHeaders(path),
    });

    return response;
  }

  async accountsCreate(holder_id, unit_symbol, holder_type) {
    let path = "/api/currencies/accounts/create/";

    let data = JSON.stringify({ holder_id: holder_id, unit_symbol: unit_symbol, holder_type: holder_type });

    let response = await this.client.post(path, data, {
      headers: await this._getHeaders(path, data),
    });

    return response;
  }

  async unitsList(filters = {}) {
    let path = "/api/currencies/units/?" + Qs.stringify(filters, { arrayFormat: "comma" });

    let response = await this.client.get(path, {
      headers: await this._getHeaders(path),
    });

    return response;
  }

  async adjustmentsList(filters = {}) {
    let path = "/api/currencies/adjustments/?" + Qs.stringify(filters, { arrayFormat: "comma" });

    let response = await this.client.get(path, {
      headers: await this._getHeaders(path),
    });

    return response;
  }

  async adjustmentsCreate(holder_id, unit_symbol, amount, description, auto_reject_timeout) {
    let path = "/api/currencies/adjustments/create/";

    let data = JSON.stringify({
      holder_id: holder_id,
      unit_symbol: unit_symbol,
      amount: amount,
      description: description,
      auto_reject_timeout: auto_reject_timeout,
    });

    let response = await this.client.post(path, data, {
      headers: await this._getHeaders(path, data),
    });

    return response;
  }

  async adjustmentsConfirm(uuid, status_description) {
    let path = "/api/currencies/adjustments/confirm/";

    let data = JSON.stringify({ uuid: uuid, status_description: status_description });

    let response = await this.client.post(path, data, {
      headers: await this._getHeaders(path, data),
    });

    return response;
  }

  async adjustmentsReject(uuid, status_description) {
    let path = "/api/currencies/adjustments/reject/";

    let data = JSON.stringify({ uuid: uuid, status_description: status_description });

    let response = await this.client.post(path, data, {
      headers: await this._getHeaders(path, data),
    });

    return response;
  }

  async transfersList(filters = {}) {
    let path = "/api/currencies/transfers/?" + Qs.stringify(filters, { arrayFormat: "comma" });

    let response = await this.client.get(path, {
      headers: await this._getHeaders(path),
    });

    return response;
  }

  async transfersCreate(from_holder_id, to_holder_id, transfer_rule, amount, description, auto_reject_timeout) {
    let path = "/api/currencies/transfers/create/";

    let data = JSON.stringify({
      from_holder_id: from_holder_id,
      to_holder_id: to_holder_id,
      transfer_rule: transfer_rule,
      amount: amount,
      description: description,
      auto_reject_timeout: auto_reject_timeout,
    });

    let response = await this.client.post(path, data, {
      headers: await this._getHeaders(path, data),
    });

    return response;
  }

  async transfersConfirm(uuid, status_description) {
    let path = "/api/currencies/transfers/confirm/";

    let data = JSON.stringify({ uuid: uuid, status_description: status_description });

    let response = await this.client.post(path, data, {
      headers: await this._getHeaders(path, data),
    });

    return response;
  }

  async transfersReject(uuid, status_description) {
    let path = "/api/currencies/transfers/reject/";

    let data = JSON.stringify({ uuid: uuid, status_description: status_description });

    let response = await this.client.post(path, data, {
      headers: await this._getHeaders(path, data),
    });

    return response;
  }

  async exchangesList(filters = {}) {
    let path = "/api/currencies/exchanges/?" + Qs.stringify(filters, { arrayFormat: "comma" });

    let response = await this.client.get(path, {
      headers: await this._getHeaders(path),
    });

    return response;
  }

  async exchangesCreate(holder_id, exchange_rule, from_unit, to_unit, from_amount, description, auto_reject_timeout) {
    let path = "/api/currencies/exchanges/create/";

    let data = JSON.stringify({
      holder_id: holder_id,
      exchange_rule: exchange_rule,
      from_unit: from_unit,
      to_unit: to_unit,
      from_amount: from_amount,
      description: description,
      auto_reject_timeout: auto_reject_timeout,
    });

    let response = await this.client.post(path, data, {
      headers: await this._getHeaders(path, data),
    });

    return response;
  }

  async exchangesConfirm(uuid, status_description) {
    let path = "/api/currencies/exchanges/confirm/";

    let data = JSON.stringify({ uuid: uuid, status_description: status_description });

    let response = await this.client.post(path, data, {
      headers: await this._getHeaders(path, data),
    });

    return response;
  }

  async exchangesReject(uuid, status_description) {
    let path = "/api/currencies/exchanges/reject/";

    let data = JSON.stringify({ uuid: uuid, status_description: status_description });

    let response = await this.client.post(path, data, {
      headers: await this._getHeaders(path, data),
    });

    return response;
  }
}

File: libs/python/gaming_billing.py
import hashlib
import hmac
import json
from datetime import datetime, timezone
from typing import Any

import aiohttp
from yarl import URL


class GamingBillingAPI:
    def __init__(
        self,
        endpoint: str,
        service_name: str,
        secret_key: str,
        service_header: str = "X-SERVICE",
        signature_header: str = "X-SIGNATURE",
        timestamp_header: str = "X-SIGNATURE-TIMESTAMP",
    ) -> None:
        self.endpoint = URL(endpoint) / "api" / "currencies"
        self.service_name = service_name
        self.service_header = service_header
        self.signature_header = signature_header
        self.timestamp_header = timestamp_header
        self.secret_key = secret_key.encode("utf-8")
        self.session: aiohttp.ClientSession | None = None

    def _compute_signature(self, data: str) -> str:
        return hmac.digest(key=self.secret_key, msg=data.encode("utf-8"), digest=hashlib.sha256).hex()

    async def _get_headers(self, path: str, data: str | None = None) -> dict[str, str]:
        timestamp = datetime.now(timezone.utc).isoformat()

        signature_data = f"{timestamp}.{path}.{data or ''}"
        signature = self._compute_signature(signature_data)

        return {
            self.service_header: self.service_name,
            self.signature_header: signature,
            self.timestamp_header: timestamp,
            "Content-Type": "application/json",
        }

    async def _request(
        self,
        session: aiohttp.ClientSession,
        method: str,
        url: URL,
        headers: dict[str, str],
        data: str | None = None,
    ) -> dict:

        async with session.request(method, url, headers=headers, data=data) as response:
            return await response.json()

    async def holders_list(
        self,
        session: aiohttp.ClientSession,
        filters: dict | None = None,
    ) -> dict:

        filters = filters or {}
        url = (self.endpoint / "holders/").with_query(filters)
        headers = await self._get_headers(url.raw_path_qs)

        return await self._request(session, "GET", url, headers)

    async def holders_detail(
        self,
        session: aiohttp.ClientSession,
        holder_id: str,
    ) -> dict:

        url = (self.endpoint / "holders" / "detail/").with_query({"holder_id": holder_id})
        headers = await self._get_headers(url.raw_path_qs)

        return await self._request(session, "GET", url, headers)

    async def holders_create(
        self,
        session: aiohttp.ClientSession,
        holder_id: str,
        holder_type: str | None = None,
        info: dict | None = None,
    ) -> dict:

        info = info or {}
        url = self.endpoint / "holders" / "create/"
        payload: dict[str, Any] = {"holder_id": holder_id}

        if holder_type is not None:
            payload["holder_type"] = holder_type

        if info is not None:
            payload["info"] = info

        json_payload = json.dumps(payload)

        headers = await self._get_headers(url.raw_path_qs, json_payload)
        return await self._request(session, "POST", url, headers, json_payload)

    async def holders_update(
        self,
        session: aiohttp.ClientSession,
        holder_id: str,
        enabled: bool | None = None,
        info: dict | None = None,
    ) -> dict:

        if enabled is None and info is None:
            raise ValueError("At least one parameter must be provided")

        url = self.endpoint / "holders" / "update/"

        payload_data: dict[str, str | int | dict] = {"holder_id": holder_id}
        if enabled is not None:
            payload_data["enabled"] = enabled
        if info is not None:
            payload_data["info"] = info

        payload = json.dumps(payload_data)
        headers = await self._get_headers(url.raw_path_qs, payload)
        return await self._request(session, "POST", url, headers, payload)

    async def accounts_list(
        self,
        session: aiohttp.ClientSession,
        filters: dict | None = None,
    ) -> dict:

        filters = filters or {}
        url = (self.endpoint / "accounts/").with_query(filters)
        headers = await self._get_headers(url.raw_path_qs)

        return await self._request(session, "GET", url, headers)

    async def accounts_detail(
        self,
        session: aiohttp.ClientSession,
        holder_id: str,
        unit_symbol: str,
        holder_type: str | None = None,
    ) -> dict:

        url = (self.endpoint / "accounts" / "detail/").with_query({"holder_id": holder_id, "unit_symbol": unit_symbol})

        if holder_type is not None:
            url = url.update_query({"holder_type": holder_type})

        headers = await self._get_headers(url.raw_path_qs)

        return await self._request(session, "GET", url, headers)

    async def accounts_create(
        self,
        session: aiohttp.ClientSession,
        holder_id: str,
        unit_symbol: str,
        holder_type: str,
    ) -> dict:

        url = self.endpoint / "accounts" / "create/"
        payload = json.dumps({"holder_id": holder_id, "unit_symbol": unit_symbol, "holder_type": holder_type})
        headers = await self._get_headers(url.raw_path_qs, payload)

        return await self._request(session, "POST", url, headers, payload)

    async def units_list(
        self,
        session: aiohttp.ClientSession,
        filters: dict | None = None,
    ) -> dict:

        filters = filters or {}
        url = (self.endpoint / "units/").with_query(filters)
        headers = await self._get_headers(url.raw_path_qs)

        return await self._request(session, "GET", url, headers)

    async def adjustments_list(
        self,
        session: aiohttp.ClientSession,
        filters: dict | None = None,
    ) -> dict:

        filters = filters or {}
        url = (self.endpoint / "adjustments/").with_query(filters)
        headers = await self._get_headers(url.raw_path_qs)

        return await self._request(session, "GET", url, headers)

    async def adjustments_create(
        self,
        session: aiohttp.ClientSession,
        holder_id: str,
        unit_symbol: str,
        amount: float,
        description: str,
        auto_reject_timeout: int,
    ) -> dict:

        url = self.endpoint / "adjustments" / "create/"
        payload = json.dumps(
            {
                "holder_id": holder_id,
                "unit_symbol": unit_symbol,
                "amount": amount,
                "description": description,
                "auto_reject_timeout": auto_reject_timeout,
            }
        )
        headers = await self._get_headers(url.raw_path_qs, payload)

        return await self._request(session, "POST", url, headers, payload)

    async def adjustments_confirm(
        self,
        session: aiohttp.ClientSession,
        uuid: str,
        status_description: str,
    ) -> dict:

        url = self.endpoint / "adjustments" / "confirm/"
        payload = json.dumps({"uuid": uuid, "status_description": status_description})
        headers = await self._get_headers(url.raw_path_qs, payload)

        return await self._request(session, "POST", url, headers, payload)

    async def adjustments_reject(
        self,
        session: aiohttp.ClientSession,
        uuid: str,
        status_description: str,
    ) -> dict:

        url = self.endpoint / "adjustments" / "reject/"
        payload = json.dumps({"uuid": uuid, "status_description": status_description})
        headers = await self._get_headers(url.raw_path_qs, payload)

        return await self._request(session, "POST", url, headers, payload)

    async def transfers_list(
        self,
        session: aiohttp.ClientSession,
        filters: dict | None = None,
    ) -> dict:

        filters = filters or {}
        url = (self.endpoint / "transfers/").with_query(filters)
        headers = await self._get_headers(url.raw_path_qs)

        return await self._request(session, "GET", url, headers)

    async def transfers_create(
        self,
        session: aiohttp.ClientSession,
        from_holder_id: str,
        to_holder_id: str,
        transfer_rule: str,
        amount: float,
        description: str,
        auto_reject_timeout: int,
    ) -> dict:

        url = self.endpoint / "transfers" / "create/"
        payload = json.dumps(
            {
                "from_holder_id": from_holder_id,
                "to_holder_id": to_holder_id,
                "transfer_rule": transfer_rule,
                "amount": amount,
                "description": description,
                "auto_reject_timeout": auto_reject_timeout,
            }
        )
        headers = await self._get_headers(url.raw_path_qs, payload)

        return await self._request(session, "POST", url, headers, payload)

    async def transfers_confirm(
        self,
        session: aiohttp.ClientSession,
        uuid: str,
        status_description: str,
    ) -> dict:

        url = self.endpoint / "transfers" / "confirm/"
        payload = json.dumps({"uuid": uuid, "status_description": status_description})
        headers = await self._get_headers(url.raw_path_qs, payload)

        return await self._request(session, "POST", url, headers, payload)

    async def transfers_reject(
        self,
        session: aiohttp.ClientSession,
        uuid: str,
        status_description: str,
    ) -> dict:

        url = self.endpoint / "transfers" / "reject/"
        payload = json.dumps({"uuid": uuid, "status_description": status_description})
        headers = await self._get_headers(url.raw_path_qs, payload)

        return await self._request(session, "POST", url, headers, payload)

    async def exchanges_list(
        self,
        session: aiohttp.ClientSession,
        filters: dict | None = None,
    ) -> dict:

        filters = filters or {}
        url = (self.endpoint / "exchanges/").with_query(filters)
        headers = await self._get_headers(url.raw_path_qs)

        return await self._request(session, "GET", url, headers)

    async def exchanges_create(
        self,
        session: aiohttp.ClientSession,
        holder_id: str,
        exchange_rule: str,
        from_unit: str,
        to_unit: str,
        from_amount: float,
        description: str,
        auto_reject_timeout: int,
    ) -> dict:

        url = self.endpoint / "exchanges" / "create/"
        payload = json.dumps(
            {
                "holder_id": holder_id,
                "exchange_rule": exchange_rule,
                "from_unit": from_unit,
                "to_unit": to_unit,
                "from_amount": from_amount,
                "description": description,
                "auto_reject_timeout": auto_reject_timeout,
            }
        )
        headers = await self._get_headers(url.raw_path_qs, payload)

        return await self._request(session, "POST", url, headers, payload)

    async def exchanges_confirm(
        self,
        session: aiohttp.ClientSession,
        uuid: str,
        status_description: str,
    ) -> dict:

        url = self.endpoint / "exchanges" / "confirm/"
        payload = json.dumps({"uuid": uuid, "status_description": status_description})
        headers = await self._get_headers(url.raw_path_qs, payload)

        return await self._request(session, "POST", url, headers, payload)

    async def exchanges_reject(
        self,
        session: aiohttp.ClientSession,
        uuid: str,
        status_description: str,
    ) -> dict:

        url = self.endpoint / "exchanges" / "reject/"
        payload = json.dumps({"uuid": uuid, "status_description": status_description})
        headers = await self._get_headers(url.raw_path_qs, payload)

        return await self._request(session, "POST", url, headers, payload)
